File: lcm-java/lcm/lcm/TCPService.java
package lcm.lcm;

import java.net.*;
import java.io.*;
import java.util.*;
import java.util.concurrent.locks.*;
import java.util.regex.*;
import java.nio.*;

public class TCPService
{
    ServerSocket serverSocket;

    AcceptThread acceptThread;
    ArrayList<ClientThread> clients = new ArrayList<ClientThread>();
    ReadWriteLock clients_lock = new ReentrantReadWriteLock();

    int bytesCount = 0;

    public TCPService(int port) throws IOException
    {
        serverSocket = new ServerSocket(port);
        //	sock.setReuseAddress(true);
        //	sock.setLoopbackMode(false); // true *disables* loopback

        acceptThread = new AcceptThread();
        acceptThread.start();

        long inittime = System.currentTimeMillis();
        long starttime = System.currentTimeMillis();
        while (!Thread.interrupted()) {
            try {
                Thread.sleep(1000);
            } catch (InterruptedException ex) {
                break;
            }
            long endtime = System.currentTimeMillis();
            double dt = (endtime - starttime) / 1000.0;
            starttime = endtime;
            System.out.printf("%10.3f : %10.1f kB/s, %d clients\n",(endtime - inittime)/1000.0, bytesCount/1024.0/dt, clients.size());
            bytesCount = 0;
        }
        // interrupt signal received
        closeResources();
    }

    private void closeResources() throws IOException {
        acceptThread.interrupt();
        serverSocket.close();
        synchronized(clients) {
            for (ClientThread clientThread : clients) {
                clientThread.closeResources();
            }
        }
    }


    public void relay(byte channel[], byte data[])
    {
        // synchronously send to all clients.
        String chanstr = new String(channel);
        try {
            clients_lock.readLock().lock();
            for (ClientThread client : clients) {
                client.send(chanstr, channel, data);
            }
        } finally {
            clients_lock.readLock().unlock();
        }
    }

    class AcceptThread extends Thread
    {
        public void run()
        {
            while (!Thread.interrupted()) {
                try {
                    Socket clientSock = serverSocket.accept();

                    ClientThread client = new ClientThread(clientSock);
                    client.start();

                    try {
                        clients_lock.writeLock().lock();
                        clients.add(client);
                    } finally {
                        clients_lock.writeLock().unlock();
                    }
                } catch (IOException ex) {
                }
            }
        }
    }

    class ClientThread extends Thread
    {
        Socket sock;
        DataInputStream ins;
        DataOutputStream outs;

        class SubscriptionRecord
        {
            String  regex;
            Pattern pat;
            SubscriptionRecord(String regex) {
                this.regex = regex;
                this.pat = Pattern.compile(regex);
            }
        }

        ArrayList<SubscriptionRecord> subscriptions = new ArrayList<SubscriptionRecord>();
        ReadWriteLock subscriptions_lock = new ReentrantReadWriteLock();

        public ClientThread(Socket sock) throws IOException
        {
            this.sock = sock;

            ins = new DataInputStream(sock.getInputStream());
            outs = new DataOutputStream(sock.getOutputStream());

            outs.writeInt(TCPProvider.MAGIC_SERVER);
            outs.writeInt(TCPProvider.VERSION);
        }

        public void run()
        {
            ///////////////////////
            // read messages until something bad happens.
            try {
                while (true) {
                    int type = ins.readInt();
                    if (type == TCPProvider.MESSAGE_TYPE_PUBLISH) {
                        int channellen = ins.readInt();
                        byte channel[] = new byte[channellen];
                        ins.readFully(channel);

                        int datalen = ins.readInt();
                        byte data[] = new byte[datalen];
                        ins.readFully(data);

                        TCPService.this.relay(channel, data);

                        bytesCount += channellen + datalen + 8;
                    } else if(type == TCPProvider.MESSAGE_TYPE_SUBSCRIBE) {
                        int channellen = ins.readInt();
                        byte channel[] = new byte[channellen];
                        ins.readFully(channel);
                        try {
                            subscriptions_lock.writeLock().lock();
                            subscriptions.add(new SubscriptionRecord(new String(channel)));
                        } finally {
                            subscriptions_lock.writeLock().unlock();
                        }
                    } else if(type == TCPProvider.MESSAGE_TYPE_UNSUBSCRIBE) {
                        int channellen = ins.readInt();
                        byte channel[] = new byte[channellen];
                        ins.readFully(channel);
                        String re = new String(channel);
                        try {
                            subscriptions_lock.writeLock().lock();
                            for(int i=0, n=subscriptions.size(); i<n; i++) {
                                if(subscriptions.get(i).regex.equals(re)) {
                                    subscriptions.remove(i);
                                    break;
                                }
                            }
                        } finally {
                            subscriptions_lock.writeLock().unlock();
                        }
                    }
                }
            } catch (IOException ex) {
            }

            ///////////////////////
            // Something bad happened, close this connection.
            try {
                closeResources();
            } catch (IOException ex) {
            }

            try {
                clients_lock.writeLock().lock();
                clients.remove(this);
            } finally {
                clients_lock.writeLock().unlock();
            }
        }

        public void closeResources() throws IOException {
            sock.close();
        }

        public void send(String chanstr, byte channel[], byte data[])
        {
            try {
                subscriptions_lock.readLock().lock();
                for(SubscriptionRecord sr : subscriptions) {
                    if(sr.pat.matcher(chanstr).matches()) {
                        synchronized(outs) {
                            outs.writeInt(TCPProvider.MESSAGE_TYPE_PUBLISH);
                            outs.writeInt(channel.length);
                            outs.write(channel);
                            outs.writeInt(data.length);
                            outs.write(data);
                            outs.flush();
                            return;
                        }
                    }
                }
            } catch (IOException ex) {
            } finally {
                subscriptions_lock.readLock().unlock();
            }
        }
    }

    public static void main(String args[])
    {
        try {
            int port = 7700;
            if (args.length > 0)
                port = Integer.parseInt(args[0]);
            new TCPService(port);
        } catch (IOException ex) {
            System.out.println("Ex: "+ex);
        }
    }
}

File: lcm-dotnet/lcm/lcm/TCPService.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Sockets;
using System.Text.RegularExpressions;
using System.Threading;

namespace LCM.LCM
{
	public class TCPService
	{
        private TcpListener serverSocket;

        private Thread acceptThread;
        private List<ClientThread> clients = new List<ClientThread>();

        private int bytesCount = 0;
		
        /// <summary>
        /// Constructor of the TCP service object
        /// </summary>
        /// <param name="port">TCP port number</param>
		public TCPService(int port)
		{
			TcpListener tempTCPListener;
			tempTCPListener = new TcpListener(System.Net.Dns.GetHostAddresses(System.Net.Dns.GetHostName())[0], port);
			tempTCPListener.Start();
			serverSocket = tempTCPListener;
            // serverSocket.setReuseAddress(true);
            // serverSocket.setLoopbackMode(false); // true *disables* loopback
			
			acceptThread = new Thread(AcceptThreadRun);
            acceptThread.IsBackground = true;
			acceptThread.Start();
			
			long inittime = DateTime.Now.Ticks / 10000;
			long starttime = DateTime.Now.Ticks / 10000;

			while (true)
			{
				try
				{
					System.Threading.Thread.Sleep(1000);
				}
				catch (System.Threading.ThreadInterruptedException)
				{
				}

				long endtime = System.DateTime.Now.Ticks / 10000;
				double dt = (endtime - starttime) / 1000.0;
				starttime = endtime;

				Console.WriteLine("{0,10:N} : {1,10:N} kB/s, {2:D} clients", (endtime - inittime) / 1000.0, bytesCount / 1024.0 / dt, clients.Count);
				bytesCount = 0;
			}
		}
		
        /// <summary>
        /// Synchronously send a message to all clients
        /// </summary>
        /// <param name="channel">channel name</param>
        /// <param name="data">data to be relayed</param>
		public void Relay(byte[] channel, byte[] data)
        {
            string chanstr = System.Text.Encoding.GetEncoding("US-ASCII").GetString(channel);

			lock (clients)
			{
				foreach (ClientThread client in clients)
				{
					client.Send(chanstr, channel, data);
				}
			}
		}
		
		private void AcceptThreadRun()
		{
			while (true)
			{
				try
				{
					System.Net.Sockets.TcpClient clientSock = serverSocket.AcceptTcpClient();
					
					ClientThread client = new ClientThread(this, clientSock);
					client.Start();
					
					lock (clients)
					{
						clients.Add(client);
					}
				}
				catch (IOException)
				{
				}
			}
		}

        private class ClientThread
        {
            private TCPService service;
            private TcpClient sock;
            private BinaryReader ins;
            private BinaryWriter outs;
            private Thread thread;

            private class SubscriptionRecord
            {
                internal string regex;
                internal Regex pat;

                public SubscriptionRecord(string regex)
                {
                    this.regex = regex;
                    this.pat = new Regex(regex);
                }
            }

            List<SubscriptionRecord> subscriptions = new List<SubscriptionRecord>();

            public ClientThread(TCPService service, TcpClient sock)
            {
                this.service = service;
                this.sock = sock;

                ins = new BinaryReader(sock.GetStream());
                outs = new BinaryWriter(sock.GetStream());

                outs.Write(TCPProvider.MAGIC_SERVER);
                outs.Write(TCPProvider.VERSION);
            }

            public void Start()
            {
                if (thread == null)
                {
                    thread = new Thread(Run);
                    thread.IsBackground = true;
                    thread.Start();
                }
            }

            public void Run()
            {
                // read messages until something bad happens.
                try
                {
                    while (true)
                    {
                        int type = ins.ReadInt32();
                        if (type == TCPProvider.MESSAGE_TYPE_PUBLISH)
                        {
                            int channellen = ins.ReadInt32();
                            byte[] channel = new byte[channellen];
                            ReadInput(ins.BaseStream, channel, 0, channel.Length);

                            int datalen = ins.ReadInt32();
                            byte[] data = new byte[datalen];
                            ReadInput(ins.BaseStream, data, 0, data.Length);

                            service.Relay(channel, data);

                            service.bytesCount += channellen + datalen + 8;
                        }
                        else if (type == TCPProvider.MESSAGE_TYPE_SUBSCRIBE)
                        {
                            int channellen = ins.ReadInt32();
                            byte[] channel = new byte[channellen];
                            ReadInput(ins.BaseStream, channel, 0, channel.Length);

                            lock (subscriptions)
                            {
                                subscriptions.Add(new SubscriptionRecord(
                                    System.Text.Encoding.GetEncoding("US-ASCII").GetString(channel)));
                            }
                        }
                        else if (type == TCPProvider.MESSAGE_TYPE_UNSUBSCRIBE)
                        {
                            int channellen = ins.ReadInt32();
                            byte[] channel = new byte[channellen];
                            ReadInput(ins.BaseStream, channel, 0, channel.Length);

                            string re = System.Text.Encoding.GetEncoding("US-ASCII").GetString(channel);
                            lock (subscriptions)
                            {
                                for (int i = 0, n = subscriptions.Count; i < n; i++)
                                {
                                    if (subscriptions[i].pat.IsMatch(re))
                                    {
                                        subscriptions.RemoveAt(i);
                                        break;
                                    }
                                }
                            }
                        }
                    }
                }
                catch (IOException)
                {
                }

                // Something bad happened, close this connection.
                try
                {
                    sock.Close();
                }
                catch (IOException)
                {
                }

                lock (service.clients)
                {
                    service.clients.Remove(this);
                }
            }

            public virtual void Send(string chanstr, byte[] channel, byte[] data)
            {
                try
                {
                    lock (subscriptions)
                    {
                        foreach (SubscriptionRecord sr in subscriptions)
                        {
                            if (sr.pat.IsMatch(chanstr))
                            {
                                outs.Write(TCPProvider.MESSAGE_TYPE_PUBLISH);
                                outs.Write(channel.Length);
                                outs.Write(channel);
                                outs.Write(data.Length);
                                outs.Write(data);
                                outs.Flush();

                                return;
                            }
                        }
                    }
                }
                catch (IOException)
                {
                }
            }

            /******************************* Helper methods *******************************/

	        /// <summary>Reads a number of characters from the current source Stream and writes the data to the target array at the specified index.</summary>
	        /// <param name="sourceStream">The source Stream to read from.</param>
	        /// <param name="target">Contains the array of characteres read from the source Stream.</param>
	        /// <param name="start">The starting index of the target array.</param>
	        /// <param name="count">The maximum number of characters to read from the source Stream.</param>
	        /// <returns>The number of characters read. The number will be less than or equal to count depending on the data available in the source Stream. Returns -1 if the end of the stream is reached.</returns>
	        private static int ReadInput(Stream sourceStream, byte[] target, int start, int count)
	        {
		        // Returns 0 bytes if not enough space in target
                if (target.Length == 0)
                {
                    return 0;
                }

		        byte[] receiver = new byte[target.Length];
		        int bytesRead = sourceStream.Read(receiver, start, count);

		        // Returns -1 if EOF
                if (bytesRead == 0)
                {
                    return -1;
                }

                for (int i = start; i < start + bytesRead; i++)
                {
                    target[i] = (byte) receiver[i];
                }
                        
		        return bytesRead;
	        }
        }
	}
}